### PR TITLE
Refactor force browser close on panic

### PR DESCRIPTION
**Pull request by ankur22**
We're currently storing the browser pid in the context. This is an issue because:

1. When working on https://github.com/grafana/xk6-browser/pull/744 it wasn't cleanly possible to panic at the mapping layer and kill all the browser processes since the pid was in a child context which we didn't easily have access to.
2. If a user launches more than one browser, and a panic occurs, only the last launched browser will be closed since that's the only pid in the context.

This change adds a `register` which is internally updated by `BrowserProcess`. When there is a need to panic (with `k6ext.Panic`) we can use the `register`'s `ForceProcessShutdown`, which will loop over all the registered `pid`s and force kill the processes.

### CODE REVIEW

File: browserprocess/browser_process.go [hunk 1]
```diff
@@ -69,13 +69,6 @@ func New(
 	}
 
 	go func() {
-		defer func() {
```
Comment (ankur22) [inline browserprocess/browser_process.go]: Unregister doesn't give us anything, so it's been removed.

File: browserprocess/register_test.go [hunk 2]
```diff
@@ -1,57 +0,0 @@
-package browserprocess
```
Comment (ankur22) [inline browserprocess/register_test.go]: This became a bit meaningless after removing `unregister`. Testing this is a little tricky, but let me try again 🤔

File: browserprocess/browser_process.go [hunk 3]
```diff
@@ -1,4 +1,6 @@
-package common
+// Package browserprocess is in charge of the interaction
+// with the OS process that is running the browser.
+package browserprocess
```
Comment (inancgumus) [inline browserprocess/browser_process.go:3]: One suggestion: `browserprocess.BrowserProcess` stutters. Would it be better to rename it to `osext`? `osext.Process`?

That would also be consistent with the comment you wrote before the package definition: _"the interaction with the OS process"_. In the future, we can add more operating-system-related stuff into the same package without creating many packages. That is how it's done in the Go standard library as well. [See](https://pkg.go.dev/os).

**Review by inancgumus — COMMENTED** [review]
Nice idea 👍  I'll take a look at this in detail later. Here's one suggestion after skimming the PR.

File: browserprocess/register.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+	"github.com/grafana/xk6-browser/log"
+)
+
+type processState struct {
+	pid int
+}
+
+var (
+	browserProcessRegister   = map[string]*processState{} //nolint:gochecknoglobals
```
Comment (ka3de) [inline browserprocess/register.go:18]: Can we simplify this map now that we don't need to store the `killed` status as `map[string]int` being: `pid+iID: pid` and get rid of the `processState` struct?

File: browserprocess/register.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+// xk6-browser is having to shutdown due to an
+// internal error (and therefore a panic).
+func ForceProcessShutdown(ctx context.Context) {
+	browserProcessRegisterMu.Lock()
+	defer browserProcessRegisterMu.Unlock()
+
+	iID := GetIterationID(ctx)
+
+	for k, v := range browserProcessRegister {
+		if iID != "" && !strings.Contains(k, iID) {
```
Comment (ka3de) [inline browserprocess/register.go:44]: I'm a little bit skeptical about generating the keys for the register as a concatenation of `iterationID` and `pid`, and also about using `strings.Contains` to verify. I was experimenting with using to different data structures for the map from `iID->pid` and the list of `pid` but was resulting rather complex, I will try to provide an alternative. Did you consider other options?

**Review by ka3de — COMMENTED** [review]
Thanks for working on this @ankur22 ! Added a couple of comments.

**Review by ka3de — COMMENTED** [review]
I am providing a suggestion with which I was experimenting in order to avoid the usage of string concatenation for `iterationID`-`pid` relationship. Feel free to discard it or discuss just some options of it. We take advantage of the fact that when not using `test_browser`, we'll have one "default iteration ID" which is a void string. Or we could even fix that in the `GetIterationID` method.
Let me know if I misunderstood something.

File: browserprocess/register.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	"github.com/grafana/xk6-browser/log"
+)
+
+type processState struct {
+	pid int
+}
+
+var (
+	browserProcessRegister   = map[string]*processState{} //nolint:gochecknoglobals
+	browserProcessRegisterMu = sync.Mutex{}               //nolint:gochecknoglobals
```
[the remark this answers is not in the document]
Comment (ankur22) [reply browserprocess/register.go:19]: Yeah, i did consider that, but the string concatenation was a lot easier to chuck in. I'll change it to this, thanks for the suggestion 🙂

File: browserprocess/register.go (hunk 4 above)
Comment (ankur22) [reply browserprocess/register.go:18]: Good point, although it's needed in the [next PR](https://github.com/grafana/xk6-browser/pull/762) that i've created 😅 Maybe i'll revert it here still to show that it's not needed for this particular change 👍

File: browserprocess/register.go (hunk 5 above)
Comment (ankur22) [reply browserprocess/register.go:44]: Yes, it's a smell. I like the approach you suggested 👍

File: browserprocess/browser_process.go (hunk 3 above)
Comment (ankur22) [reply browserprocess/browser_process.go]: Yeah, nice idea. I've amended it in a1f72e472bcca9d4522a99892e4498da5111f771.

File: tests/test_browser.go [hunk 7]
```diff
@@ -82,6 +84,16 @@ func newTestBrowser(tb testing.TB, opts ...any) *testBrowser {
 		vu.CtxField = ctx
 	}
 
+	// This is used to help the browserprocess register
+	// distinguish between different error runs so that
+	// a panic in one test doesn't force close the browser
+	// in other running tests.
+	u, err := uuid.NewV4()
```
Comment (ankur22) [inline tests/test_browser.go:91]: I could use a global atomic counter instead of `uuid` so that we're not having to depend on another library, WDYT?

Comment (ka3de) [reply tests/test_browser.go:91]: I think the `uuid` library will be required anyway because it's included in k6, but I think an atomic counter will make it simpler so +1 for me.

File: osext/browser_process.go [hunk 8]
```diff
@@ -23,20 +25,23 @@ type BrowserProcess struct {
...
+	UserDataDir *storage.Dir
 
 	logger *log.Logger
 }
 
-func NewBrowserProcess(
+// New creates a new osext. It will start
+// a child process where the internet browser will
+// run.
+func New(
```
Comment (ka3de) [inline osext/browser_process.go:44]: Considering that we have now renamed the pkg to `osext`, should we rename this to `NewBrowserProcess`?

File: browserprocess/browser_process.go (hunk 3 above)
Comment (inancgumus) [reply browserprocess/browser_process.go]: Thanks for the amend! What about `BrowserProcess` to `Process`? I don't foresee any other process types in the near future :)

Comment (ankur22) [reply browserprocess/browser_process.go]: I like that. It's confusing otherwise. Resolved in 8af5625586f2ccd7a1a2f29451840fe955076f24

File: osext/browser_process.go (hunk 8 above)
Comment (ankur22) [reply osext/browser_process.go]: Yep, good spot. Resolved in 2cdcab34d24729f1385b9a6660f27232204ab13e

**Comment by ka3de** [comment]
Just one last comment briefly discussed in https://github.com/grafana/xk6-browser/pull/755#pullrequestreview-1288724078 about the "default iteration ID" that we can make it more explicit for example by defining a constant in `osext/context.go` such as:
```
    defIterationID = "default" // Or something similar
```
And then modify the `GetIterationID` method in order to return the default value in case the given context does not contain a value for iteration ID key. But happy to leave it as is counting on the void string as default.

**Review by ka3de — APPROVED** [review]
LGTM.

**Comment by ankur22** [comment]
> defIterationID = "default"

This is a good idea! Avoids possible confusion. Keep comments and suggestions coming if you think it could help 🙂 

Resolved in ed5498c9cb40e37b9e96e2611152e2faa4410319

**Review by ka3de — APPROVED** [review]
LGTM :tada:

**Review by inancgumus — COMMENTED** [review]
LGTM 👍 Some questions and suggestions.

I also wonder about the reason of associating processIDs to runIDs? Could we store the process IDs in `processRegister map[]int`? In the end, we want to kill processes, and I wrap my head around why do we need to associate the process IDs with run IDs? 🤔😅

**Comment by ankur22** [comment]
> I also wonder about the reason for associating processIDs with runIDs? Could we store the process IDs in `processRegister map[string]int` or just `[]int`? In the end, we want to kill processes, and I can't wrap my head around why we need to associate the process IDs with run IDs? 🤔😅 If that's possible, then maybe we can get rid of storing data in context.

For the integration tests it makes sense to work with the pid as the `runID`, since each test should be running its own browser and any panic/kill should be isolated to that test and browser instance.

When running `k6-browser` tests, the user could create two `browser` class instances, opening two separate browsers:

```javascript
  const browser = chromium.launch({headless: false});
  const context = browser.newContext();
  const page = context.newPage();

  const browser2 = chromium.launch({headless: false});
  const context2 = browser2.newContext();
  const page2 = context2.newPage();
```

If we were to panic and kill based on the `pid`, we'd only force close one browser, and not the other. This is why we have a `default` `runId`.

> Nit: I also prefer to use something other than package-level variables. Although more straightforward at the beginning, they are error-prone and can be overwritten by different parts of code. Do you think we should find another way?

Yes, i've been thinking about this too. I'm not a fan of doing it this way, but as far as I can tell there's no nice way to inject a `register` instance to both the `osext.Process` (which is registering the browser pids) and `k6ext.Panic` without affecting the current API and a lot of code.

The only way I could think of was with the context, but i think that a misuse of the context (although we're already doing it with `vu`).

Eventually though, we will be able to find a clean way when we get to the point of only panicking in the `mapping` layer, at that point it should be easy to inject an instance of `register` into the mapping layer as well into `osext.Process`.

File: tests/test_browser.go [hunk 9]
```diff
@@ -35,6 +37,8 @@ type testBrowser struct {
 	api.Browser
 }
 
+var iID uint64 //nolint:gochecknoglobals
```
[the remark this answers is not in the document]
Comment (ankur22) [reply tests/test_browser.go:40]: I think the responsibility of the `runID` `uint64` and where it's incremented should be done in the tests. The whole `runID`"hack" is there due to the tests, so if we can keep as much of it with the `test` package then I think that's a better signal to future devs.

**Comment by ankur22** [comment]
@inancgumus @ka3de take a look at https://github.com/grafana/xk6-browser/pull/762, which could affect this PR too 😅

**Review by inancgumus — CHANGES_REQUESTED** [review]
Please see #779.
